### PR TITLE
Refactor raw pointers in QMCDrivers into unique_ptr to address leak

### PR DESCRIPTION
**Pull request by williamfgc**
## Proposed changes

Refactor raw pointers in QMCDrivers into unique _ptr for RAII (automatic deallocation).
Affects OperatorBase, addOperator and makeclone, getHamiltonian functions that retrieve/interact with OperatorBase
Raw pointers were creating a leak as part of #3083 identified tests
Tried unique_ptr first, but these pointers have shared ownership (some tests failed).
Must still use pointers due to runtime polymorphism.
Exploring `unique_ptr` as opposed to `shared_ptr` as discussed in PR #3126

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- Yes: internal API changed for OperatorBase, no external change

## What systems has this change been tested on?
Ubuntu 20.04, not all tests currently passing. Trying to address in this WIP PR.

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCDrivers/WFOpt/HamiltonianRef.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  {
+    const auto LocalEnergyComponent = Hrefs_[i].get().evaluate(P);
+    if (std::isnan(LocalEnergyComponent))
+      APP_ABORT("QMCHamiltonian::evaluate component " + Hrefs_[i].get().myName + " returns NaN\n");
+    LocalEnergy += LocalEnergyComponent;
+  }
+  return LocalEnergy;
+}
+
+int HamiltonianRef::addObservables(ParticleSet& P)
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/HamiltonianRef.cpp:54]: this one is only used by the legacy CUDA build.

**Comment by williamfgc** [comment]
@ye-luo thanks for making this move forward by fixing the issues on my branch. I see the same pesky seg fault I can't reproduce as in #3126. Any ideas?

**Comment by ye-luo** [comment]
Test this please

File: src/QMCHamiltonians/MPC.cpp [hunk 2]
```diff
@@ -30,18 +30,12 @@ namespace qmcplusplus
...
 }
 
-MPC::~MPC()
-{
-  if (VlongSpline)
-    destroy_Bspline(VlongSpline);
-  if (DensitySpline)
-    destroy_Bspline(DensitySpline);
-}
+MPC::~MPC() = default;
```
Comment (williamfgc) [inline src/QMCHamiltonians/MPC.cpp:38]: Nice!

**Comment by williamfgc** [comment]
@ye-luo great catch! Thanks! I will close PR #3126 in favor of this. One more leak down.

**Comment by ye-luo** [comment]
I did non-trivial changes in in the optimization routines. So probably it is better to have a review from @markdewing

**Comment by ye-luo** [comment]
@markdewing get a chance to review this? Or we can merge this now and you do post-merge review when you have spare cycles?

**Comment by ye-luo** [comment]
> Looks okay. Keeping the list of Hamiltonian components as strings rather than converting to objects as quickly as possible is odd, but I think it works.

The names are used as "references". I wanted to directly copy H to h0_ptr_list_ directly but the tricked by the internal Observables. For that reason, I used an intermediate H_KE which I hope can be remove soon.

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**